consensus/votecounter/round_data.go

```diff
@@ -0,0 +1,80 @@
+package votecounter
+
+import (
+	"github.com/NethermindEth/juno/consensus/types"
+	"github.com/NethermindEth/juno/utils"
+)
+
+type roundMap[V types.Hashable[H], H types.Hash, A types.Addr] = map[types.Round]*roundData[V, H, A]
+
+type roundData[V types.Hashable[H], H types.Hash, A types.Addr] struct {
+	proposal               *types.Proposal[V, H, A]
+	uncountedProposerPower types.VotingPower
+	perIDVotes             map[H]*ballotSet[A]
+	nilVotes               ballotSet[A]
+	allVotes               ballotSet[A]
+}
+
+func newRoundData[V types.Hashable[H], H types.Hash, A types.Addr]() roundData[V, H, A] {
+	return roundData[V, H, A]{
+		proposal:               nil,
+		uncountedProposerPower: 0,
+		perIDVotes:             make(map[H]*ballotSet[A]),
+		nilVotes:               newBallotSet[A](),
+		allVotes:               newBallotSet[A](),
+	}
+}
+
+func (r *roundData[V, H, A]) setProposal(proposal *types.Proposal[V, H, A], votingPower types.VotingPower) bool {
+	if r.proposal != nil {
+		return false
+	}
+	r.proposal = proposal
+
+	proposerBallot, ok := r.allVotes.ballots[proposal.Sender]
+	if proposerNotVoted := !ok || (!proposerBallot[Prevote] && !proposerBallot[Precommit]); proposerNotVoted {
+		r.uncountedProposerPower = votingPower
+	}
+	return true
+}
+
+func (r *roundData[V, H, A]) addVote(vote *types.Vote[H, A], votingPower types.VotingPower, voteType VoteType) bool {
+	var perVote *ballotSet[A]
+	if vote.ID != nil {
+		if perVote = r.perIDVotes[*vote.ID]; perVote == nil {
+			perVote = utils.HeapPtr(newBallotSet[A]())
+			r.perIDVotes[*vote.ID] = perVote
```

Comment:
What's the rationale behind storing a pointer to `perVote` instead of `perVote` itself? You can still keep the pointer, assuming not a reference on the map, you can do:

```golang
r.perIDVotes[*vote.ID] = newBallotSet[A]()
perVote = &r.perIDVotes[*vote.ID]
```

Reply:
I don't think the second line is possible, as map is not addressable.